invenio_vocabularies/contrib/funders/schema.py

```diff
@@ -8,17 +8,95 @@
 
 """Funders schema."""
 
+from functools import partial
+
 from flask_babelex import lazy_gettext as _
-from marshmallow import Schema, fields, validate
-from marshmallow_utils.fields import SanitizedUnicode
+from marshmallow import Schema, ValidationError, fields, post_load, pre_dump, \
+    pre_load, validate, validates_schema
+from marshmallow_utils.fields import IdentifierSet, SanitizedUnicode
+from marshmallow_utils.schemas import IdentifierSchema
+
+from ...services.schema import BaseVocabularySchema
+from .config import funder_schemes
 
 
 class FunderRelationSchema(Schema):
     """Funder schema."""
 
+    name = SanitizedUnicode(
+        validate=validate.Length(min=1, error=_('Name cannot be blank.'))
+    )
+    id = SanitizedUnicode()
+
+    @validates_schema
+    def validate_funder(self, data, **kwargs):
+        """Validates that either id either name are present."""
+        id_ = data.get("id")
+        name = data.get("name")
+        if id_:
+            data = {"id": id_}
+        elif name:
+            data = {"name": name}
+
+        if not id_ and not name:
+            raise ValidationError(
+                _("An existing id or a free text name must be present"),
+                "affiliations"
+            )
+
+
+class FunderSchema(BaseVocabularySchema):
+    """Service schema for funders."""
+
+    pid = SanitizedUnicode(required=True)
     name = SanitizedUnicode(
         required=True,
         validate=validate.Length(min=1, error=_('Name cannot be blank.'))
     )
-    scheme = SanitizedUnicode()
-    identifier = SanitizedUnicode()
+    country = SanitizedUnicode()
+    identifiers = IdentifierSet(fields.Nested(
+        partial(
+            IdentifierSchema,
+            allowed_schemes=funder_schemes,
+            identifier_required=False
+        )
+    ))
+
+    # TODO: the 3 hooks below could be extracted to a PIDField
```

Comment:
@slint Should we create said field? I do not recall what did you said on the call this morning sorry

Reply:
resolving this comment because it boils down to the one below: i.e. where do we managed the pid/id? in the component or in marshmallow